homeassistant/components/switch/arest.py

```diff
@@ -68,6 +70,69 @@ def is_on(self):
         """ True if device is on. """
         return self._state
 
+
+class ArestSwitchFunction(ArestSwitchBase):
+    """ Implements an aREST switch. Based on functions. """
+
+    def __init__(self, resource, location, name, func):
+        super().__init__(resource, location, name)
+        self._func = func
+
+        request = requests.get('{}/{}'.format(self._resource, self._func),
+                               timeout=10)
+
+        if request.status_code is not 200:
+            _LOGGER.error("Can't find function. Is device offline?")
+            return
+
+        try:
+            request.json()['return_value']
+        except KeyError:
+            _LOGGER.error("Function doesn't include return value. Is the function name correct.")
+        except ValueError:
+            _LOGGER.error("Function response invalid. Is the function name correct.")
+
+    def turn_on(self, **kwargs):
+        """ Turn the device on. """
+        request = requests.get('{}/{}'.format(self._resource, self._func),
+                               timeout=10, params={"params": "1"})
+
+        if request.status_code == 200:
+            self._state = True
+        else:
+            _LOGGER.error("Can't turn on function %s at %s. Is device offline?",
+                          self._func, self._resource)
+
+    def turn_off(self, **kwargs):
+        """ Turn the device off. """
+        request = requests.get('{}/{}'.format(self._resource, self._func),
+                               timeout=10, params={"params": "0"})
+
+        if request.status_code == 200:
+            self._state = False
+        else:
+            _LOGGER.error("Can't turn off function %s at %s. Is device offline?",
+                          self._func, self._resource)
+
+    def update(self):
```

Comment:
I'm not sure if this is better, More things would have to parametrize the "url path", the "params" and the "log message" too, even the `_pin` and `_func` attribute would need to be aligned.

It feels a bit like to much generalization too me. But I can do it if you want.